<a href="https://colab.research.google.com/github/SOUMEE2000/Sentiment-Analysis-guidelines-IMDB-Datset-/blob/main/1.%20Using_different_vectorisers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**SKlearn has a variety of different hash vectorisers and it can be interesting to know how even things such vectoriser, stemming and lemmastisation can affect the the accuracy even though they are not directly related to the model**

In [ ]:
!unzip IMDB.zip

Archive:  IMDB.zip
  inflating: IMDB Dataset.csv        


In [ ]:
import pandas as pd
df= pd.read_csv('IMDB Dataset.csv')

In [ ]:
df.head()

review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

In [ ]:
# make entire text lowercase
df['review_processed'] = [review.lower() for review in df['review']]

# Replacing short words
df['review_processed'] = df['review_processed'].apply(lambda x: ' '.join([w for w in x.split() if len(w)>2]))


In [ ]:
# Removing Stopwords Begin
import nltk
nltk.download('punkt')
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk import word_tokenize
stop_words = stopwords.words('english')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
# Making custom list of words to be removed 
add_words = ['movie','film','one','make','even','the']

In [ ]:
stop_words.extend(add_words)

In [ ]:
# Function to remove stop words 
def remove_stopwords(rev):
    review_tokenized = word_tokenize(rev)
    rev_new = " ".join([i for i in review_tokenized  if i not in stop_words])
    return rev_new

# Removing stopwords
df['review_processed'] = [remove_stopwords(r) for r in df['review_processed']]

In [ ]:
feature=[]
for i in df["sentiment"]:
    if i== "positive":
        feature.append(1)
    elif i =="negative":          # if "negative" was not specified, length of df wouldn't match with len(features)
        feature.append(0)

In [ ]:
df["feature"]= feature

In [ ]:
# Begin Lemmatization 
"""from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
nltk.download('wordnet') # Run atleast once 
nltk.download('averaged_perceptron_tagger')
# function to convert nltk tag to wordnet tag
lemmatizer = WordNetLemmatizer()

# Finds the part of speech tag
def nltk_tag_to_wordnet_tag(nltk_tag):
    if nltk_tag.startswith('J'):
        return wordnet.ADJ
    elif nltk_tag.startswith('V'):
        return wordnet.VERB
    elif nltk_tag.startswith('N'):
        return wordnet.NOUN
    elif nltk_tag.startswith('R'):
        return wordnet.ADV
    else:          
        return None

# lemmatize sentence using pos tag
def lemmatize_sentence(sentence):
    #tokenize the sentence and find the POS tag for each token
    nltk_tagged = nltk.pos_tag(nltk.word_tokenize(sentence))  
    #tuple of (token, wordnet_tag)
    wordnet_tagged = map(lambda x: (x[0], nltk_tag_to_wordnet_tag(x[1])), nltk_tagged)
    lemmatized_sentence = []
    for word, tag in wordnet_tagged:
        if tag is None:
            #if there is no available tag, append the token as is
            lemmatized_sentence.append(word)
        else:        
            #else use the tag to lemmatize the token
            lemmatized_sentence.append(lemmatizer.lemmatize(word, tag))
    return " ".join(lemmatized_sentence)


df['review_processed'] = df['review_processed'].apply(lambda x: lemmatize_sentence(x))"""

'from nltk.stem import WordNetLemmatizer\nfrom nltk.corpus import wordnet\nnltk.download(\'wordnet\') # Run atleast once \nnltk.download(\'averaged_perceptron_tagger\')\n# function to convert nltk tag to wordnet tag\nlemmatizer = WordNetLemmatizer()\n\n# Finds the part of speech tag\ndef nltk_tag_to_wordnet_tag(nltk_tag):\n    if nltk_tag.startswith(\'J\'):\n        return wordnet.ADJ\n    elif nltk_tag.startswith(\'V\'):\n        return wordnet.VERB\n    elif nltk_tag.startswith(\'N\'):\n        return wordnet.NOUN\n    elif nltk_tag.startswith(\'R\'):\n        return wordnet.ADV\n    else:          \n        return None\n\n# lemmatize sentence using pos tag\ndef lemmatize_sentence(sentence):\n    #tokenize the sentence and find the POS tag for each token\n    nltk_tagged = nltk.pos_tag(nltk.word_tokenize(sentence))  \n    #tuple of (token, wordnet_tag)\n    wordnet_tagged = map(lambda x: (x[0], nltk_tag_to_wordnet_tag(x[1])), nltk_tagged)\n    lemmatized_sentence = []\n    for word, 

In [ ]:
df.head()

review  ... feature
0  One of the other reviewers has mentioned that ...  ...       1
1  A wonderful little production. <br /><br />The...  ...       1
2  I thought this was a wonderful way to spend ti...  ...       1
3  Basically there's a family where a little boy ...  ...       0
4  Petter Mattei's "Love in the Time of Money" is...  ...       1

[5 rows x 4 columns]

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer= TfidfVectorizer(max_features=2500)
vectorizer.fit(df.review_processed)

TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.float64'>, encoding='utf-8',
                input='content', lowercase=True, max_df=1.0, max_features=2500,
                min_df=1, ngram_range=(1, 1), norm='l2', preprocessor=None,
                smooth_idf=True, stop_words=None, strip_accents=None,
                sublinear_tf=False, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, use_idf=True, vocabulary=None)

In [ ]:
X = vectorizer.transform(df.review_processed).toarray()

In [ ]:
y= df.feature.values

In [ ]:
#from sklearn.preprocessing import MinMaxScaler
#scaling = MinMaxScaler(feature_range=(0,1)).fit(X)
#X = scaling.transform(X)

In [ ]:
X

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [ ]:
# Splitting the dataset into train and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

In [ ]:
import numpy as np
from sklearn.naive_bayes import MultinomialNB
classifier= MultinomialNB()
classifier.fit(X_train, y_train)
## Testing the model on test set
y_pred = classifier.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
accuracy = accuracy_score(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)
print(cm)
print("The model accuracy is", accuracy )

[[4230  805]
 [ 698 4267]]
The model accuracy is 0.8497


# **Some Results**

 **Hash Vectoriser**
 
 * naive_bayes(GNB) = 70.93%
 * naive_bayes(MNB),(CNB) somehow negative values are being passed
 * To get around that used the MinMax scaler CNB= 77.5%, MNB=76.8%

**Tfidf vectorizer**
* GNB= 81.01%   81.75%(after properly removing stopwords)
* MNB= 84.69%   84.97%(,,)
* CNB= 84.7%    85%(,,)

**Without stopwords lemmatisation takes longer, CNB= 84.22%, GNB= 80.52%, MNB= 84.27%(surprisingly)**

**Without lemmatisation, MNB= 84.42%, CNB= 84.46%, GNB=81.92%**

**With optimisation GNB= 81.92%**